File: source/c/arena_allocator.c
// gcc -o arena_allocator arena_allocator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ARENA_SIZE 1024

char arena[ARENA_SIZE];
int offset = 0;

void *arena_alloc(int size) {
    if (offset + size > ARENA_SIZE) {
        printf("arena out of memory!\n");
        return NULL;
    }
    void *p = arena + offset;
    offset += size;
    return p;
}

void arena_reset(void) { offset = 0; }

int main(void) {
    int *a = arena_alloc(4);
    int *b = arena_alloc(4);
    int *c = arena_alloc(4);
    int *d = arena_alloc(4);
    int *e = arena_alloc(4);

    *a = 100; *b = 200; *c = 300; *d = 400; *e = 500;

    printf("slot 0: %d\n", *a);
    printf("slot 1: %d\n", *b);
    printf("slot 2: %d\n", *c);
    printf("slot 3: %d\n", *d);
    printf("slot 4: %d\n", *e);

    printf("arena used: %d bytes\n", offset);
    // Stack-allocated arena, no explicit free needed
    printf("arena freed\n");
    return 0;
}

File: source/rust/arena_allocator.rs
use std::alloc::{alloc, dealloc, Layout};

struct Arena {
    data: *mut u8,
    offset: usize,
    size: usize,
}

impl Arena {
    fn new(size: usize) -> Self {
        let layout = Layout::from_size_align(size, 8).unwrap();
        let data = unsafe { alloc(layout) };
        Arena { data, offset: 0, size }
    }

    fn alloc(&mut self, bytes: usize) -> Option<*mut u8> {
        if self.offset + bytes > self.size {
            println!("arena out of memory!");
            return None;
        }
        let p = unsafe { self.data.add(self.offset) };
        self.offset += bytes;
        Some(p)
    }

    fn reset(&mut self) { self.offset = 0; }
}

impl Drop for Arena {
    fn drop(&mut self) {
        unsafe { dealloc(self.data, Layout::from_size_align(self.size, 8).unwrap()); }
    }
}

fn main() {
    let mut arena = Arena::new(1024);
    let slots: Vec<*mut u8> = (0..5).map(|_| arena.alloc(4).unwrap()).collect();
    let vals = [100i32, 200, 300, 400, 500];
    for (i, &slot) in slots.iter().enumerate() {
        unsafe { *(slot as *mut i32) = vals[i]; }
    }
    for (i, &slot) in slots.iter().enumerate() {
        let v = unsafe { *(slot as *const i32) };
        println!("slot {}: {}", i, v);
    }
    println!("arena used: {} bytes", arena.offset);
    drop(arena);
    println!("arena freed");
}
